src/sol_binder/contracts/instance.py

```diff
@@ -139,46 +141,41 @@ def transact(self, func_name: str, func_args, tx_args: TxParams = defaultdict())
         """
         if not tx_args.get('from'):
             tx_args['from'] = self.__default_account
-        func: ContractFunction = cast(ContractFunction, self._contract.functions[func_name])
-        nonce = self.__get_nonce_for_transact(tx_args)
-        tx_args.update({"nonce": nonce, })
-
-        try:
-            tx = func(*func_args).buildTransaction(tx_args)
-        except Exception as e:
-            raise TransactionBuildError(e)
+        if tx_args.get('nonce'):
+            @contextmanager
+            def get_nonce(account):
```

Comment:
Consider that instead of creating an anonymous function in order to supply get_nonce, in the case of pre-specified nonce one could simply skip the with statement altogether and proceed to do the TX

This also provides a good way to look at the breakdown of the code: 
(WARNING: Pseudo-correct code)
```
def transact(self, ...):
    ...
    is_prespecified_nonce = 'nonce' in tx_args
    if is_prespecified_none:
        __internal_transact(tx_args, ...)
    else:
        with get_nonce(...):
             __internal_transact(tx_args, ...)

def __internal_transact(self, tx_args, nonce):
  tx_args.update({"nonce": nonce, })
  try:
      tx = func(*func_args).buildTransaction(tx_args)
  except Exception as e:
  ...... # more code of transact
```

Reply:
In addition, since nonce is considered to be managed - I'm not sure we should allow manual nonce
For tests we can create a "test nonce manager"

And even if a manual nonce needs to be specified, it may be more logical to override the NonceManager rather than actually specify it for transact.

Please consider and choose what looks best to you

Reply:
As agreed, lets not accept any manual nonce and throw a ManualNonceNotSupported("Please read about 'Nonce Managers' for solbinder")